Jenkinsfile: reduce CI time by disabling windows RS1

tiborvass opened the pull request.

andrewhsu committed:
skip win-RS1 on PRs unless the checkbox is checked

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>
Signed-off-by: Tibor Vass <tibor@docker.com>

andrewhsu commented:
This PR is copied from #39899 

Regarding issue #39532, the `windowsRS1` PR checks are well over the 60-minute time limit desired for PRs (#39651) by being the longest pole at 100 minutes. Additionally, the `windowsRS1` are consistently flaky on various tests: 10 out of the [last 19 PRs merged this week](https://github.com/moby/moby/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aclosed+merged%3A2019-09-04..2019-09-10) have `windowsRS1` failures.

We still have `windowsRS5` PR checks and I suggest we open a separate PR to reduce flakiness and runtime of `windowsRS1` PR checks after this one is merged.

cc @StefanScherer @vikramhh @thaJeztah @ddebroy

StefanScherer commented:
@andrewhsu 👍 Makes sense to work on windowsRS1 in a separate PR like parallelizing tests, remove or fix flaky tests, ...

thaJeztah commented on Jenkinsfile:
```diff
@@ -697,9 +697,10 @@ pipeline {
                     }
                 }
-                stage('win-RS1') {
+                stage('win-RS1-master') {
```
don't think we have to rename the stage

Do we know if Jenkins has a param that says "it's a PR"? I think we want to run these on release branches as well, so basically what we want to filter on is; "don't run this on PR jobs"

thaJeztah replied:
Looks like there is; https://stackoverflow.com/a/48103602/1811501

(and then remove `branch 'master'` from the example below)

```groovy
        when {
            allOf {
                environment name: 'CHANGE_ID', value: ''
                branch 'master'
            }
        }
```

thaJeztah replied:
And the `allOf` looks not needed; https://jenkins.io/doc/book/pipeline/syntax/

> If the `when` directive contains more than one condition, all the child conditions must return true for the stage to execute. This is the same as if the child conditions were nested in an `allOf` condition

thaJeztah replied:
oh; this is more elegant; and the official way; https://jenkins.io/doc/book/pipeline/syntax/

```groovy
when {
    not { changeRequest() }
}
```

andrewhsu replied:
i've updated

thaJeztah commented on Jenkinsfile:
```diff
@@ -700,6 +700,7 @@ pipeline {
                 stage('win-RS1') {
                     when {
                         beforeAgent true
+                        not { changeRequest() }
```
👍 guess we should update the other `master` selectors the same.

thaJeztah replied:
Actually; just thinking of a slightly improved approach;

change `params.windowsRS1` default to `false`, and then do this;

```groovy
or {
  not { changeRequest() }
  expression { params.windowsRS1 }
}
```

If that works;

- RS1 (and, e.g. `ppc64le/s390x-integration-cli`) only run on non-PR jobs ("master")
- Unless someone ticks the `windowsRS1` checkbox

That way, we can easily run these tests on a specific PR if we see a need

andrewhsu committed:
run integration-cli stages on s390x and ppc64le if not a PR check

Essentially, run on merge to target branch which may or may not be
master branch. Could be 19.03 branch, for example.

See: https://jenkins.io/doc/book/pipeline/syntax/

Signed-off-by: Andrew Hsu <andrewhsu@docker.com>

andrewhsu commented:
@tiborvass could you update the title of this PR? I believe this is good to go now cc @thaJeztah 

[Failure](https://ci.docker.com/public/job/moby/job/PR-39789/18/execution/node/210/log/?consoleFull) is unrelated:
```
00:49:50.929      --- FAIL: Test/DockerSuite/TestRunInteractiveWithRestartPolicy (11.61s)
00:49:50.929          docker_cli_run_test.go:1800: assertion failed: 
00:49:50.929              Command:  /usr/local/cli/docker run -i --name test-inter-restart --restart=always busybox sh
00:49:50.929              ExitCode: 0
00:49:50.929              Stdout:   
00:49:50.929              Stderr:   
00:49:50.929              
00:49:50.929              Failures:
00:49:50.929              ExitCode was 0 expected 11
```

tiborvass commented:
@andrewhsu feel free to tell me a better title

thaJeztah commented on Jenkinsfile:
```diff
@@ -728,7 +728,11 @@ pipeline {
                         beforeAgent true
-                        expression { params.windowsRS1 }
+                        // Skip this stage on PRs unless the windowsRS1 checkbox is selected
+                        anyOf {
```
can you make this same change to the other stages we disable on PR's?

andrewhsu replied:
I'm not sure that would make sense for the existing `s390x` and `s390x integration-cli` stages without a new build parameter. Could be an idea for a future PR.

For this PR, I'd prefer to keep the scope to reducing CI times by disabling windows RS1.

thaJeztah replied:
oh 🤦‍♂ you're right, I thought we had a parameter for those, but `ppc64le` is for the stage that's enabled by default

thaJeztah commented:
Failure on amd64 is unrelated; https://ci.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-39789/runs/18/nodes/209/log/?start=0

Should've been fixed by https://github.com/moby/moby/pull/39853, but looks like it's still flaky 😢 /cc @tao12345666333



```
[2019-09-13T16:59:55.446Z]     --- FAIL: Test/DockerSuite/TestRunInteractiveWithRestartPolicy (11.61s)
[2019-09-13T16:59:55.446Z]         docker_cli_run_test.go:1800: assertion failed: 
[2019-09-13T16:59:55.446Z]             Command:  /usr/local/cli/docker run -i --name test-inter-restart --restart=always busybox sh
[2019-09-13T16:59:55.446Z]             ExitCode: 0
[2019-09-13T16:59:55.446Z]             Stdout:   
[2019-09-13T16:59:55.446Z]             Stderr:   
[2019-09-13T16:59:55.446Z]             
[2019-09-13T16:59:55.446Z]             Failures:
[2019-09-13T16:59:55.446Z]             ExitCode was 0 expected 11
```

thaJeztah reviewed (approved):
LGTM

tao12345666333 commented:
@thaJeztah Thanks for ping, I will continue to try to fix it.
